File: scripts/train-ml-policy-service.mjs
import readline from "readline";
import { runMlPolicyTraining } from "../lib/ml-policy-train-runner.mjs";

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
  terminal: false
});

function respond(payload) {
  process.stdout.write(`${JSON.stringify(payload)}\n`);
}

respond({ type: "ready", service: "train-ml-policy-js" });

rl.on("line", (line) => {
  const trimmed = String(line || "").trim();
  if (!trimmed) return;
  let msg;
  try {
    msg = JSON.parse(trimmed);
  } catch {
    respond({ type: "error", error: "invalid_json" });
    return;
  }

  if (msg?.action === "shutdown") {
    respond({ type: "bye" });
    process.exit(0);
    return;
  }

  if (msg?.action !== "train") {
    respond({ id: msg?.id ?? null, type: "error", error: "unknown_action" });
    return;
  }

  const t0 = Date.now();
  try {
    const res = runMlPolicyTraining({
      style: String(msg?.style || ""),
      feedbackFile: String(msg?.feedbackFile || ""),
      outputModel: String(msg?.outputModel || ""),
      summaryFile: String(msg?.summaryFile || ""),
      useCache: msg?.useCache !== false
    });
    respond({
      id: msg?.id ?? null,
      type: "result",
      elapsedMs: Date.now() - t0,
      ...res
    });
  } catch (err) {
    respond({
      id: msg?.id ?? null,
      type: "error",
      elapsedMs: Date.now() - t0,
      error: err?.message || String(err)
    });
  }
});

File: scripts_py/train_ml_policy_service.py
import json
import sys
import time

from train_ml_policy import run_training


def emit(payload):
    sys.stdout.write(json.dumps(payload, ensure_ascii=False) + "\n")
    sys.stdout.flush()


def main():
    emit({"type": "ready", "service": "train-ml-policy-py"})
    for line in sys.stdin:
        raw = (line or "").strip()
        if not raw:
            continue
        try:
            msg = json.loads(raw)
        except Exception:
            emit({"type": "error", "error": "invalid_json"})
            continue

        if msg.get("action") == "shutdown":
            emit({"type": "bye"})
            return
        if msg.get("action") != "train":
            emit({"id": msg.get("id"), "type": "error", "error": "unknown_action"})
            continue

        t0 = time.perf_counter()
        try:
            res = run_training(
                feedback_file=str(msg.get("feedbackFile") or "data/training/feedback.ndjson"),
                output_model=str(msg.get("outputModel") or "models/prosody-policy-v1.json"),
                summary_file=str(msg.get("summaryFile") or ""),
                style_scope=str(msg.get("style") or ""),
                ridge=float(msg.get("ridge", 1e-6)),
                use_cache=bool(msg.get("useCache", True)),
            )
            emit(
                {
                    "id": msg.get("id"),
                    "type": "result",
                    "elapsedMs": round((time.perf_counter() - t0) * 1000, 3),
                    **res,
                }
            )
        except Exception as err:
            emit(
                {
                    "id": msg.get("id"),
                    "type": "error",
                    "elapsedMs": round((time.perf_counter() - t0) * 1000, 3),
                    "error": str(err),
                }
            )


if __name__ == "__main__":
    main()
